nova-tt/pause.hpp

```diff
@@ -21,33 +21,44 @@
 #ifndef NOVA_TT_PAUSE_HPP
 #define NOVA_TT_PAUSE_HPP
 
-#ifdef __SSE2__
-#include <emmintrin.h>
+#if defined(__i386__) || defined(_M_IX86) || \
+    defined(__x86_64__) || defined(_M_X64)
+/* Intel */
+#define NOVA_TT_PAUSE
+#include <immintrin.h>
+#elif (defined(__ARM_ARCH_6K__) || \
+       defined(__ARM_ARCH_6Z__) || \
+       defined(__ARM_ARCH_6ZK__) || \
+       defined(__ARM_ARCH_6T2__) || \
+       defined(__ARM_ARCH_7__) || \
+       defined(__ARM_ARCH_7A__) || \
+       defined(__ARM_ARCH_7R__) || \
+       defined(__ARM_ARCH_7M__) || \
+       defined(__ARM_ARCH_7S__) || \
+       defined(__ARM_ARCH_8A__) || \
+       defined(__aarch64__))
+/* ARM */
+/* mnemonic 'yield' is supported from ARMv6k onwards */
+#define NOVA_TT_YIELD
+#else
+/* fallback */
+#warning "architecture does not support yield/pause instruction; falling back to thread yielding."
+#include <thread>
 #endif
 
 namespace nova   {
 namespace detail {
 
-#ifdef __SSE2__
-
 static inline void pause()
 {
+#if defined(NOVA_TT_PAUSE)
 	_mm_pause();
-}
-
-#elif defined(__GNUC__) && ( defined(__i386__) || defined(__x86_64__) )
-
-static inline void pause()
-{
-	__asm__ __volatile__( "rep; nop" : : : "memory" );
-}
-
-#else
-
-static inline void pause()
-{}
-
+#elif defined(NOVA_TT_YIELD)
+	__asm__ __volatile__("yield");
```

Comment:
it seems that `isb` is the better approach here, as `yield` can boil down to a `nop`: https://stackoverflow.com/questions/70810121/why-does-hintspin-loop-use-isb-on-aarch64

Reply:
Interesting! I didn't know about `isb`. Definitely worth considering. Will try to push an update later.